KAFKA-7026: Sticky Assignor Partition Assignment Improvement (KIP-341)

vahidhashemian opened the pull request:
In the current implementation of sticky assignor the leader does not cache the most recent calculated assignment. It relies on the fact that each consumer in the group sends its subscribed topics and also its current assignment when a rebalance occurs. This could lead to the issue described in [KAFKA-7026](https://issues.apache.org/jira/browse/KAFKA-7026), in which current assignment of a consumer is no longer valid and should be ignored. The solution implemented in this PR involves the leader caching the most recent assignment of each consumer, so the assignment reported by a consumer can be properly validated (and ignored if necessary) by the leader during a rebalance.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vahidhashemian committed:
KAFKA-7026: Sticky Assignor Partition Assignment Improvement

In the current implementation of sticky assignor the leader does not cache the most recent calculated assignment.
It relies on the fact that each consumer in the group sends its subscribed topics and also its current assignment when a rebalance occurs.
This could lead to the issue described in KAFKA-7026, in which current assignment of a consumer is no loonger valid and should be ignored.
The solution implemented in this PR involves the leader caching the most recent assignment of each consumer, so the assignment reported by a consumer can be properly ignored if necessary.

steven-aerts commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,17 +277,36 @@
+                    }
+                }
+
+                currentAssignment.put(member, deserializedAssignment);
```
This caching logic might work if the leader does not change.

But I have the impression that it is possible to end up in the same problem when the leader changes.  Because as far as I see, the new leader will not have build up its cache.  (or even worse has an outdated cache from a previous time it was leader)

Which I think can still cause troubles.

I see two other possible solutions:
- The easiest one to implement I think is keep the previous code, but change the type of the `currentAssignment` parameter from a `Map<String, List<TopicPartition>>` to a `Map<TopicPartition, String>`, or keep the type but do a post-processing step on this list where double TopicPartitions are removed before doing the `sortPartitions`.

- Another possibility is putting a generation counter in the schema, and only keep assignments with the maximum generation counter.  (or those where it is still missing for backwards compatibility).

The first possible solutions is easy to implement but makes it a little less sticky. (which I do not mind)
The second solution is more sound, but implementing it in a backwards compatible way is a challenge.

steven-aerts reviewed (changes requested):
Hello Vahid,

I have some reservations on your solution.  As I fear it will not work if the leader changes.  (or even worse when an previous leader becomes leader again).

Maybe I am wrong, but I think a different kind of solution is needed.  If you want I can take a stab at it.

tedyu commented:
I would vote for #1 above since #2 would take some time to get right.

Changing type to Map<TopicPartition, String> seems to be more efficient.

vahidhashemian commented:
@steven-aerts thanks for the feedback. The provided solution, based on some tests I ran, should resolve the bug in the use case described in the JIRA (a leader pausing for a bit, another consumer becoming a leader, and then the old leader rejoining). Can you describe more concretely a scenario that still exposes a bug in the assignor?

steven-aerts commented:
@vahidhashemian ,

my main problem with this solution is that you keep some state on the leader.  So the other consumers do not have any state which means that if any of them becomes leader, I think theoretically it is again possible to see this issue.
I think this solution reduces the change of this problem significantly, but does not take it away.
Next to that if a consumer lost its leadership, and later on is again elected as a leader, its cache will be totally out of sync which will make its next assignment nearly random instead of sticky.

## To give you a concrete scenario.
### schematic representation of the bug
Let's start with a schematic representation of the bug.
G1: c3 joins group and gets assignment:
```
c1(leader): subscription=[1,2,3] --> assignment=[1,2]
c2        : subscription=[4,5,6] --> assignment=[4,5]
c3        : subscription=null    --> assignment=[3,6]
```
G2: c3 is unreachable and leaves according to c1/the coordinator the group:
```
c1(leader): subscription=[1,2]   --> assignment=[1,2,3]
c2        : subscription=[4,5]   --> assignment=[4,5,6]
c3(gone)  : subscription=[3,6]   -->X
```
G3: c3 comes back with subscription from G1 combined with the G1 subscriptions of G2 for c1 and c2 this leads to a wrong assignment:
```
c1(leader): subscription=[1,2,3] --> assignment=[1,2,3]
c2        : subscription=[4,5,6] --> assignment=[4,5,6]
c3(back)  : subscription=[3,6]   --> assignment=[3,6]
```

### Now a case where your solution works:
G1: c3 joins group and gets assignment:
```
c1(leader): subscription=[1,2,3],cache={c1:[1,2,3], c2:[4,5,6]}     --> assignment=[1,2]
c2        : subscription=[4,5,6],cache=null                         --> assignment=[4,5]
c3        : subscription=null,cache=null                            --> assignment=[3,6]
```
G2: c3 is unreachable and leaves according to c1/the coordinator the group:
```
c1(leader): subscription=[1,2],cache={c1:[1,2], c2:[4,5], c3:[3,6]} --> assignment=[1,2,3]
c2        : subscription=[4,5],cache=null                           --> assignment=[4,5,6]
c3(gone)  : subscription=[3,6],cache=null                           -->X
```
G3: c3 comes back with subscription from G1 but cache drops its subscription data:
```
c1(leader): subscription=[1,2,3],cache={c1:[1,2,3], c2:[4,5,6]}     --> assignment=[1,2]
c2        : subscription=[4,5,6],cache=null                         --> assignment=[4,5]
c3(back)  : subscription=[3,6],cache=null                           --> assignment=[3,6]
```
### Where I think it breaks
But if in the meanwhile another consumer becomes leader you have a problem, as G3 will then look like this as the cache will not prevent the double assignment:
```
c1        : subscription=[1,2,3],cache={c1:[1,2,3], c2:[4,5,6]}     --> assignment=[1,2,3]
c2(leader): subscription=[4,5,6],cache=null                         --> assignment=[4,5,6]
c3(back)  : subscription=[3,6],cache=null                           --> assignment=[3,6]
```

So my main problem is at the end, you are still able assign a double partition, as no doubles are filtered anywhere.

vahidhashemian commented:
@steven-aerts Thanks for describing those scenarios in detail. Though, regarding the last one (where you think it breaks) I'm not sure about how the group can end up in that state. The scenario you described would imply that at some point `c1` left the group (leader change happens only when existing leader leaves the group). Therefore, at the following rebalance after `c1` leaves, `c2` will build its cache, and will not have a `null` cache when `c3` comes back above. Do you agree?

I tried to rebuild the scenario with that in mind, but couldn't reproduce the issue (see below).

G1: `c3` joins group and gets assignment:
```
c1 (leader): subscription=[1,2,3],cache={c1:[1,2,3], c2:[4,5,6]}     --> assignment=[1,2]
c2         : subscription=[4,5,6],cache=null                         --> assignment=[4,5]
c3         : subscription=null,cache=null                            --> assignment=[3,6]
```

G2: `c3` pauses
```
c1 (leader): subscription=[1,2],cache={c1:[1,2], c2:[4,5], c3:[3,6]} --> assignment=[1,2]
c2         : subscription=[4,5],cache=null                           --> assignment=[4,5]
```

G3: `c1` pauses, `c2` becomes leader
```
c2 (leader): subscription=[4,5],cache=null                           --> assignment=[1,2,3,4,5,6]
```

G4: `c1` and `c3` resume: `c2` has built the cache and would ignore invalid subscriptions by `c1` and `c3`
```
c1         : subscription=[1,2],cache={c1:[1,2], c2:[4,5], c3:[3,6]} --> assignment=[1,2]
c2 (leader): subscription=[1,2,3,4,5,6],cache={c2:[1,2,3,4,5,6]}     --> assignment=[4,5]
c3         : subscription=[3,6],cache=null                           --> assignment=[3,6]
```

steven-aerts commented:
@vahidhashemian 
I agree it is a lot more difficult to trigger as you need in 1 generation 3 operations (`c1` disappears, `c1` appears, `c3` disappears).
A scenario where you need only 2 operations is if in `G3`, `c1` disappears and `c3` appears in one step.

Then G3 should look like this where partition 6 still has a double assignment:
```
c1(gone)  : subscription=[1,2,3],cache={c1:[1,2,3], c2:[4,5,6]}     --> X
c2(leader): subscription=[4,5,6],cache=null                         --> assignment=[1,4,5,6]
c3(back)  : subscription=[3,6],cache=null                           --> assignment=[2,3,6]
```

For this to work you still two operations in one generation.
If we have somehow a guarantee that a leader in a consumer group will have operations atomically applied it might be your solution works  (although I am even not convinced of that when you add more partitions and consumers to the mix).
But as far as I can see this is not guaranteed by kafka.  

Reproducing it by hand by pausing debuggers might be tricky.  What I would do to try to reproduce the scenario above is pause the coordinator (use a kafka cluser of 1), unpause `c3`, terminate `c1` and then unpause the coordinator.  And hopefully `c2` will become leader while seeing `c3`s assignment.

Good luck,

   Steven

vahidhashemian commented:
@steven-aerts I understand. There are corner cases where the issue can resurface, even with this PR. I'll try to think about ways to improve this solution. Between the two options you suggested I like the second one best (using generations), as the first one could jeopardize stickiness (we should avoid randomly removing duplicates). I'll think about this some more. Thanks again for pointing out this issue.

steven-aerts commented:
@vahidhashemian 

I think there is a variant of option 1 which also ensures stickiness in (almost?) all cases and will be easier to implement then option two.

I think if you detect duplicates, you can loop over all subscriptions, find the subscription with the most duplicates, and remove from that subscription all duplicate partitions assignments, and test again.
Iff there are two assignments with the same number of duplicates, take the assignment with the least partitions.
With this logic I did not find any assignment where you loose stickiness while you do not need any extra meta data.

What I propose is something like this:
```java
for( ; ; ) {
    Set<TopicPartition> doubleAssigned = currentAssignment.values().stream().flatMap(List::stream)
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
            .entrySet().stream().filter(e -> e.getValue() > 1).map(Map.Entry::getKey)
            .collect(Collectors.toSet());
    if (doubleAssigned.isEmpty()) {
        break;
    }
    String toClean = currentAssignment.entrySet().stream().max(
            Comparator.comparing(e -> e.getValue().stream().filter(doubleAssigned::contains).count())
                    .thenComparing(e -> e.getValue().size(), Comparator.reverseOrder()))
            .get().getKey();

    currentAssignment.get(toClean).removeIf(doubleAssigned::contains);
}
```

This is just an example.
I did not test this at all.  It also uses a lot of java 8 constructs which will make it difficult to cherry pick it to older releases, and the comparator will be very inefficient if there are a lot of consumers as the number of duplicated partitions is then calculated over and over again.

I hope this helps.

vahidhashemian commented:
@steven-aerts If I understood your suggestion correctly, even though it will preserve stickiness in most situations, but there are still examples that it fails to do so. For example, let's assume a group of 5 consumers: `c1`, ..., `c5`. If `c2`, `c3`, `c4`, `c5` drop briefly and then rejoin the group together, `c1` will have the most duplicates, even though it has a valid (up-to-date) assignment. So invalidating `c1`'s duplicate assignments will hurt stickiness. Please advise if I misunderstood your solution.

I think with generation markers we can preserve stickiness in the safest possible way: any consumer with a stale generation marker loses its assignment in case of a duplicate. I think the code doesn't have to change much to support that.

steven-aerts commented:
@vahidhashemian 
I agree that in your example my heuristic will alter the assignment of `c1` which is the latest generation.
On the other hand, `c2`-`c5` have to receive partitions from `c1` to balance everything out so it might be better that they receive partitions they were previously assigned, instead of a random one.  Which will happen when we work with generation markers.   On the other hand this is really an extreme case.

Any solution is good for me, as long as it excludes a duplicate assignment and preserves stickiness in 95% of the cases.  So if you see a way to introduce generation counter also OK for me.

Thanks

vahidhashemian committed:
Changed the solution to use generation markers instead of caching the assignment by the leader

This helps in situations where a new leader is chosen and there are dupliacates in the reported assignments by consumers.

vahidhashemian commented:
cc @hachikuji, please review! Thanks!

steven-aerts commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,18 +303,80 @@
+        // note that a conflict could exists only if user data is for different generations
+
+        Map<TopicPartition, ConsumerGenerationPair> partitionConsumer = new HashMap<>();
+        int maxGeneration = 0;
```
Detail: just to be sure, I would initialize it to `this.generation`, to make sure the generation always increments.

vahidhashemian replied:
My thinking was that this generation marker is okay to reset to 0 in case at some point in time none of the consumers in the group carry user data, which would have the same effect even if we start with `this.generation` here (if no consumer has user data, `this.generation = 0`). I made the change as you suggested.

steven-aerts commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,18 +303,80 @@
+                    if (consumerUserData.generation == existingRecord.generation) {
+                        // same partition is assigned to two consumers during the same rebalance
+                        throw new IllegalStateException("Error: Partition '" + partition + "' is assigned to multiple consumers " +
+                                "following sticky assignment generation " + consumerUserData.generation + ".");
```
Small remark: By throwing here an exception you are giving up.
But I am wondering if you should not try to fix it.

I do not know which consistency guarantees a kafka consumer group exactly gives.  But if it is possible to get into a split brain situation (network split), where temporarily a consumer group is split in two and gets two leaders.  You get in a situation where the `SitckyAssignor` will never recover.
If it was me, I would write a big fat error, and drop one of the two assignments.

The same for line#322.

vahidhashemian replied:
That makes sense. I updated this to make sure it keeps running.

steven-aerts commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -421,10 +515,17 @@ private int getBalanceScore(Map<String, List<TopicPartition>> assignment) {
             TreeSet<String> sortedConsumers = new TreeSet<>(new SubscriptionComparator(assignments));
             sortedConsumers.addAll(assignments.keySet());
 
+            List<TopicPartition> prevPartitions = new ArrayList(prevAssignment.keySet());
```
Maybe I am wrong, but shouldn't this be moved inside the `while` loop, so it is initialized every time a new assignment is selected?
Because, now I think only the first consumer will be challenged against this previous partitions.

vahidhashemian replied:
You're right. The whole list needs to be compared against each consumer's assignment. Updated.

steven-aerts reviewed (changes requested):
Looks good.

The only place where I am doubting is the `prevPartitions` initialization.

Thanks

vahidhashemian committed:
Minor fixes

vahidhashemian reviewed (commented):
@steven-aerts thanks for taking a look and raising those issues. I tried to address them in the new commit.

steven-aerts commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,18 +303,81 @@
+            else if (consumerGeneration.generation == existingPair.generation)
+                // same partition is assigned to two consumers during the same rebalance
+                throw new IllegalStateException("Error: Partition '" + partition + "' had been assigned to " +
+                        "multiple consumers following sticky assignment generation " + consumerGeneration.generation + ".");
```
Just wondering, if this exception stays (and did not became an error log) because it is impossible to hit and should have been prevented by the previous error log?

vahidhashemian replied:
You're right. I missed replacing this `throw` statement. It's done in the new patch. Thanks!

vahidhashemian committed:
Minor update

steven-aerts reviewed (approved):
:+1: cannot wait to see this merged.

I also like the fact that this is pure java 7 so it can be backported to the bugfix releases.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -188,21 +189,45 @@
     private static final Schema STICKY_ASSIGNOR_USER_DATA = new Schema(
-            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT)));
+            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT)),
+            new Field(GENERATION_KEY_NAME, Type.INT32));
```
Have you thought about compatibility with the old schema? I am wondering if it would be better to somehow use the generation from the consumer group so that we don't have to change this schema.

vahidhashemian replied:
@hachikuji, thanks for the feedback, I had thought about it, but noticed it needed some updates. The new patch should address it.

Regarding using the existing consumer group generation, I think it may not worth the trouble and added complexity. We don't have consumer group generation on the client side (as far as I can tell), and protocol changes will be required. Plus, supporting the old schema seems to be simple enough. What do you think?

hachikuji replied:
Have you verified that this change is forward compatible as well? Older versions of the sticky assignor need to be able to work with the updated version.

The client definitely has access to the group's generation. The question is whether and how to expose it to the assigner.

vahidhashemian replied:
I tested this with a mix of old and new clients (using old and new versions of sticky assignor), and things seem to be running smoothly. The old clients deserializes only the partition array of the schema from the new clients and ignores their generation marker.

Sorry, you're right. Generation seems to be available in the coordinator. But exposing it to just the sticky assignor could add complexity (other assignors don't need it). I still think the existing solution addresses the raised concern. But if you feel that using the consumer group generation is the way to go I can explore that approach. Thanks!

steven-aerts replied:
@vahidhashemian when I first proposed the generation counters, I looked at the coordinator code, and I had the impression that you have only access to the current / your own generation counter, but not the ones used/known by the other generation counters.

So my interpretation is, that if you make this generation accessible to the `Assignor` you can only use it as an input source to for the generation counter we send in every ConsumerUserData.
So I think its value is rather limited.

vahidhashemian replied:
@steven-aerts yes, that should still work if we decide to take that path. The leader always uses the latest generation and sends that along with the assigned partitions to each consumer in the group. Then later if a consumer drops for a brief period they would have a stale assignment (with an old generation). The leader knows to ignore those partition assignments in the case of a conflict. Would that make sense?

steven-aerts replied:
It certainly makes sense, however, I do not see much added benefit compared to the more or less arbitrary generation counter you implemented in this PR.
If it was already exposed, you could use it, but I would not expose it just for this use case.

hachikuji replied:
Frankly I think you need a KIP for this change either way since you are changing the schema for the sticky assigner user data. Anything exposed on the wire is a public API. I'm actually a little skeptical that this solution can even work if you are using a local generation counter, but I have not reviewed the implementation in detail, so I could be wrong. 

The advantage of using the generation from the group is that it is guaranteed to increase monotonically after every rebalance. Exposing it in `PartitionAssignor.Assignment` seems reasonable at a glance and may be useful for other stateful assigners.

vahidhashemian replied:
I can write a short KIP. I have tested this in some common use cases and have not noticed an issue. I'll let you review this implementation, and I'll look at how your suggestion would work from an implementation point of view. Thanks!

vahidhashemian replied:
[KIP-341](https://cwiki.apache.org/confluence/display/KAFKA/KIP-341%3A+Update+Sticky+Assignor%27s+User+Data+Protocol) is created for this PR.

vahidhashemian committed:
Fixed backward schema compatibility

vahidhashemian committed:
Added a unit test to cover conflicting previous assignments

vahidhashemian committed:
Switch to using consumer group generation in user data

(instead of the locally managed generation)

kscaldef commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java:
```diff
@@ -42,16 +42,30 @@
+                                                    Map<String, Subscription> subscriptions,
+                                                    int generation) {
+        return assign(partitionsPerTopic, subscriptions);
+    }    @Override
```
There's a formatting error at this line.

ean5533 commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -421,10 +518,18 @@ private int getBalanceScore(Map<String, List<TopicPartition>> assignment) {
             TreeSet<String> sortedConsumers = new TreeSet<>(new SubscriptionComparator(assignments));
             sortedConsumers.addAll(assignments.keySet());
 
+            List<TopicPartition> prevAssignedPartitions = new ArrayList(prevAssignment.keySet());
```
This `ArrayList` should be generic (that is, `new ArrayList<>(...)`).

vahidhashemian committed:
Minor updates

kscaldef commented:
I may be missing something, but it seems to me like onAssignment() ought to be extracting the generation from the userData and updating the field value.  Otherwise, won't all the non-leader members of the consumer group keep always sending generation=0 with their current assignments?

vahidhashemian commented:
@kscaldef You are correct. Thanks for catching this issue. I'll try to fix it in the next commit.

vahidhashemian committed:
Fixed a couple of user data related issues

steven-aerts commented:
Bump, as I think this one can be merged.
Let me know if there is still something which I can do for this issue.

Thanks,


   Steven

vahidhashemian commented:
I'd be happy to rebase this PR if it can make it to 2.1.0. cc @hachikuji

kscaldef commented:
Forgive me for being ignorant about the usual process, but will this also be backported to earlier versions?  This is a pretty significant bug in this feature.

vahidhashemian commented:
retest this please

vahidhashemian commented:
@hachikuji I'd appreciate your review again when you have a chance. Unit test failures seem to be unrelated.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java:
```diff
@@ -56,12 +56,25 @@
      *         for all members who in the input subscription map.
      */
-    Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions);
+    default Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions) {
```
In a similar scenario in KIP-336, we only implemented the default behavior for the new method. I wonder if we should do the same here?

vahidhashemian replied:
Updated this in the new patch and pushed the implementation of the existing method to `AbstractPartitionAssignor`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -185,24 +185,51 @@
+    }
+
+    static final class ConsumerGenerationPair {
+        String consumer;
```
nit: seems these could be final? Same in `ConsumerUserData`.

vahidhashemian replied:
Right, updated in the new patch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+        return assignments;
+    }
+
+    public int generation() {
```
nit: I assume this is only needed for testing? Can we reduce access?

vahidhashemian replied:
Correct, I updated this in the new patch to use package access. Thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+            }
         }
+
+        ++generation;
```
Can you explain the purpose of this? I would expect we'd want to only use the generation passed from the consumer.

vahidhashemian replied:
This is where the generation is bumped up during a rebalance (a sticky partition assignment) and guarantees that the next set of assignments have a higher generation value. The added unit tests verify this. Please let me know if I missed your point.

hachikuji replied:
Why does the generation need to be bumped? We are using the value that comes from the coordinator, so doesn't that already happen for us?

vahidhashemian replied:
You're right. After looking into this a bit more carefully I realized this bump can be skipped (even though it doesn't normally lead to a double-increase).

hachikuji reviewed (commented):
@vahidhashemian Sorry for the delay. I left a few minor comments. At a higher level, I'm a little confused about the logic. I think the basic idea is that the group generation allows us to detect members which are joining with a stale previous assignment. I am wondering why we don't simply discard such assignments. In other words, if the previous assignment generation is not exactly one less than the generation passed to `assign`, then we know the assignment is invalidated by a more recent assignment and we can ignore it. It seems like the current logic attempts to do finer-grained reasoning about generations at the partition level. Is it really necessary?

vahidhashemian reviewed (commented):
@hachikuji Thanks for taking a look at this PR. I responded to your code comments inline.

Regarding the logic around keeping the previous assignment, I see two potential benefits: 1) the consumer might somehow benefit from the preserved state (despite the interrupt) that may not be possible with a fresh assignment. It's as if we're trying to preserve stickiness across generation gaps too. 2) The work required by the assignor could be reduced as the number of partitions that need to be assigned to a potential consumer will be dropped (compared to a quick validation of the previously assigned partitions).

I don't see an obvious downside in starting with the previous assignment. If I'm missing something please let me know. Thanks!

vahidhashemian committed:
Minor updates based on feedbacky

hachikuji commented:
@vahidhashemian Here is my confusion. If there is a "generation gap," then that means the consumer which did not participate in the latest generation will have had its partitions assigned to another member and it would not be able to retain them anyway. The only case in which this wouldn't happen is if the group becomes empty, but I'm not sure that's worth optimizing for.

The reason I ask about it is that it seems to add some additional complexity. The sticky assignor is already way more complicated than the other assignors, so I was seeking simplification.

vahidhashemian commented:
@hachikuji Given that the use case that leads to the issue fixed by this KIP/PR is usually a brief disconnect, the consumer that was disconnected will rejoin with the same subscription as before. Therefore, it would make sense for it to have the same set of partitions (roughly) after the rebalance. It has to get some partitions anyways, so why not give it what it had before (as much as possible)? As I mentioned, because we are pre-assigning its previous partitions back to itself (and perform a trivial validation), there are fewer undecided partitions left that need to go through the heavier process of sticky partition assignment from scratch. In my opinion, with this approach we are reducing the complexity a bit because the more the partitions we can pre-assign to consumers the lighter the work needed to find potential consumers for the partitions and balance the assignments.

If this is not helping with the confusion or if you think my argument is not valid I can start changing the implementation as you suggested :)

Thanks!

kscaldef commented:
IMO, restoring the old assignment in this situation is not clearly desirable and may be counter-productive.  In the sort of stateful systems in which we make use of the sticky assignor, we pay a non-trivial cost in state restoration and event reprocessing every time there is a reassignment.  If there are two reassignments, we pay that cost twice with no clear benefit.

hachikuji commented:
To make this concrete, let's use the following example. Say we have three consumers in a group: A, B, and C. They are consuming partitions 0-5 from a single topic. In generation 1, we have the following assignment:

A: 0, 3
B: 1, 4
C: 2, 5

Suppose that A then falls out of the group and its partitions are assigned to B and C in generation 2.

B: 0, 1, 4
C: 2, 3, 5

Now A returns to the group and claims an assignment of [0, 3] from generation 1. I think Vahid is saying that it is still beneficial in this scenario to return [0, 3] to A. I guess that seems reasonable. 

One thing I feel slightly uncomfortable about is the fact that the generation gap is not visible to the user in any way. This means it is not safe for a rebalance listener to assume that assigned partitions weren't temporarily assigned elsewhere, so it must still do some state/offset initialization. From the perspective of a rebalance listener, the user would see [0, 3] assigned (in generation 1) then [0, 3] assigned again (in generation 3). The only way I can think to address this would be to expose the generation to the rebalance listener. This seems like an orthogonal problem though. 

It's worth mentioning that KIP-345 would make situations like this uncommon. If using "static membership," A will generally not fall out of the group.

vahidhashemian commented:
Thanks @hachikuji for the clarification and the example provided. I agree with you completely.

With respect to generation gap issue we can think about how to improve the issues you brought up as a follow up task.

Also, even though with KIP-345 this issue may not easily surface again, this PR could still help with older versions and can be applied as a hot fix (of course that's on top of making the logic more sound).

Please advise if you think there are further improvements to be made here. Thanks again!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+                    } else if (consumerUserData.generation > existingRecord.generation) {
+                        // same partition is assigned to two consumers in different rebalances.
+                        // the assignment being processed has higher generation than the one already processed, and takes priority
+                        updatePrevAssignment(prevAssignment, partition, existingRecord);
```
It may make this less confusing if you explain the difference between `partitionConsumer` and `prevAssignment`. Perhaps the former needs a better name?

vahidhashemian replied:
Added a description and renamed `partitionConsumer` to `currentPartitionConsumer`.

vahidhashemian committed:
Removed the generation bump in StickyAssignor

vahidhashemian reviewed (commented):
@hachikuji I updated the PR based on your feedback and removed the redundant generation bump as it wasn't necessary. Thanks!

vahidhashemian committed:
Simplify finding current and last consumer of each partition by using a sorted map

vahidhashemian committed:
Made improvements based on the feedback

vahidhashemian committed:
Pass on the generation from the coordinator only after assignment is complete

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+            else if (consumerGeneration.generation == existingPair.generation) {
+                // same partition is assigned to two consumers during the same rebalance.
+                // log an error and skip this record
+                log.error("Partition '" + partition + "' had been assigned to multiple consumers " +
```
nit: Use `{}` substitutions. Maybe this could be `warn` since there's nothing the user can actually do about it shouldn't actually cause any problems for the current assignment.

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+        // note that a conflict could exists only if user data is for different generations
+
+        // currentPartitionConsumer will hold the latest ConsumerGenerationPair associated with a partition
+        // while prevAssignment holds the prior ConsumerGenerationPair (before current) of each parition
```
typo: `parition`

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,115 @@
+            if (userData == null || !userData.hasRemaining()) continue;
+            ConsumerUserData consumerUserData = deserializeTopicPartitionAssignment(userData);
+
+            for (TopicPartition partition: consumerUserData.partitions) {
```
Effectively what we are doing is sorting the assignments for each partition using the generation and picking the latest two for the current and previous assignments. Is that right? Could we simplify the logic by building a `SortedMap<Integer, String>` for each partition where the key is the generation and the value is the memberID? Then the current assignment would be the last entry and the previous assignment would be the one prior.

vahidhashemian replied:
Great suggestion! It helps simplifying the code and makes `currentPartitionConsumer` redundant. I tried refactoring the code to accommodate this in the new patch.

hachikuji reviewed (commented):
Thanks for the explanation. Added a couple more comments.

vahidhashemian reviewed (commented):
Thanks @hachikuji for taking another look at this. I submitted a new patch, and also responded to your comments inline.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +306,102 @@
+        // note that a conflict could exists only if user data is for different generations
+
+        // for each partition we create a sorted map of its consumers by generation
+        Map<TopicPartition, Map<Integer, String>> sortedPartitionConsumersByGeneration = new HashMap<>();
```
We could use `SortedMap` (or even `TreeMap`) here instead of the generic `Map`. Then we wouldn't need the ugly cast below.

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +306,102 @@
+            for (TopicPartition partition: consumerUserData.partitions) {
+                if (sortedPartitionConsumersByGeneration.containsKey(partition)) {
+                    Map<Integer, String> consumers = sortedPartitionConsumersByGeneration.get(partition);
+                    if (consumers.containsKey(consumerUserData.generation)) {
```
One scenario where we may hit this is during an upgrade when we have some consumers on older versions. Might not be worth checking that though.

vahidhashemian replied:
Are you referring to when `generation == -1`? If so, do you think we need to handle that scenario differently?

hachikuji replied:
I was not sure. It seems like having a generation=-1 doesn't cause any spoilers here, but I'm always suspicious when sentinels get mixed into the code. One way to force ourselves to be honest would be to use `Optional<Integer>`. Maybe it's good enough if we just have a good test case.

vahidhashemian replied:
I switched to using `Optional<Integer>` for `ConsumerUserData.generation`as you suggested.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +306,102 @@
+
+            // let's process the current (most recent) consumer first
+            String consumer = consumers.get(it.next());
+            if (currentAssignment.containsKey(consumer))
```
You can probably simplify this using `computeIfAbsent`.

vahidhashemian replied:
👍  I hope I got it right :)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +306,102 @@
+                int generation = it.next();
+                prevAssignment.put(partition, new ConsumerGenerationPair(consumers.get(generation), generation));
+            }
+            it.remove();
```
This seems unnecessary since we're throwing away the collection anyway.

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java:
```diff
@@ -42,8 +42,24 @@
      */
-    public abstract Map<String, List<TopicPartition>> assign(Map<String, Integer> partitionsPerTopic,
-                                                             Map<String, Subscription> subscriptions);
+    public Map<String, List<TopicPartition>> assign(Map<String, Integer> partitionsPerTopic,
```
It might be safer to leave this abstract. Otherwise we should choose a better sentinel for the generation. Perhaps we should change the API below to use `Optional<Integer>` instead?

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -400,7 +400,7 @@ public long timeToNextPoll(long now) {
         log.debug("Performing assignment using strategy {} with subscriptions {}", assignor.name(), subscriptions);
 
-        Map<String, Assignment> assignment = assignor.assign(metadata.fetch(), subscriptions);
+        Map<String, Assignment> assignment = assignor.assign(metadata.fetch(), subscriptions, latestGeneration().generationId);
```
This might be a little brittle since it depends on the `generation` field getting updated before the rebalance has completed. Perhaps we can pass `generation` as an argument to `performAssignment`?

vahidhashemian replied:
That makes sense. Added the `generation` argument. Thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -519,14 +616,15 @@ private boolean canParticipateInReassignment(String consumer,
                          Map<TopicPartition, String> currentPartitionConsumer) {
         boolean initializing = currentAssignment.get(sortedCurrentSubscriptions.last()).isEmpty();
-        boolean reassignmentPerformed = false;
+        boolean reassignmentPerformed; // starts with a 'false' default value
```
nit: the comment seems more verbose than defining the default directly.

vahidhashemian replied:
Fair enough. I changed it back.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +306,102 @@
     }
 
+    @Override
+    protected Map<String, Assignment> getAssignments(Map<String, List<TopicPartition>> rawAssignments) {
```
I am a bit confused about this. Previously we weren't defining user data for the assignment. We only needed to provide the old assignment in the subscription user data. Was that not sufficient?

vahidhashemian replied:
I believe it has not been sufficient because we have been sending assignments without user data to coordinator, and have been getting them back without them too. I'll look more closely at the existing code to see if I'm missing something. The comment in `AbstractPartitionAssignor` kind of indicates this too:

> // this class maintains no user data, so just wrap the results

hachikuji reviewed (commented):
Thanks for the updates. Just a few more comments.

vahidhashemian reviewed (commented):
Thanks @hachikuji. Submitted a patch to address your suggestions.

vahidhashemian commented:
Unit test failures are related to the method signature change in the latest commit. Will fix them in the next commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java:
```diff
@@ -42,8 +43,24 @@
-                                                             Map<String, Subscription> subscriptions);
+    public Map<String, List<TopicPartition>> assign(Map<String, Integer> partitionsPerTopic,
+                                                    Map<String, Subscription> subscriptions) {
+        return assign(partitionsPerTopic, subscriptions, Optional.of(0));
```
`Optional.empty`?

vahidhashemian replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java:
```diff
@@ -58,10 +58,21 @@
+     * @return A map from the members to their respective assignment. This should have one entry
+     *         for all members who in the input subscription map.
+     */
+    default Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions, int generation) {
```
Hmm.. Now that I'm thinking about it, this seems like the wrong place to pass through generation. We do not need to know the current generation in order to sort the current and previous assignments. We should only need to know the current generation when we receive the assignment after the rebalance completes. I think we should have the following method instead:

```java
void onAssignment(Assignment assignment, Optional<Integer> generation);
```

Perhaps this is the cause of the confusion over propagation of the generation through the Assignment UserData.

vahidhashemian replied:
Hmm, would this work? The reason we introduced the `generation` argument in `assign` method was to pass on the generation value from the coordinator to the sticky assignor so we can perform the assignment by comparing the consumer's reported generation/assignment with the current assignment. Passing it on in `onAssignment` only solves one half of the problem (consumers knowing what their assignment generation is). It, however, doesn't help the assignor know the current assignment. Are you implying we don't care what the current assignment is when performing the sticky assignment and just rely on the sorted generation assignments we built?

hachikuji replied:
Yes, I'm suggesting that the current generation is not needed when performing the assignment. All we need to know is the generation of the previous assignment reported by each member, which is included in the subscription user data. Does that make sense? Then we get all the information we need for the next rebalance inside `onAssignment`.

vahidhashemian replied:
Ok. Thanks for clarifying. That should work. I'll work on modifying the code accordingly later today.

vahidhashemian reviewed (commented):
@hachikuji Thanks for the recent comments. I'll submit another patch soon, just first need to get a clarification on your comment inline.

vahidhashemian commented:
@hachikuji I refactored the code to feed in generation to StickyAssignor after assignment is complete.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java:
```diff
@@ -58,12 +59,12 @@
+     * @param generation The consumer group generation associated with this partition assignment (optional)
      */
-    void onAssignment(Assignment assignment);
+    void onAssignment(Assignment assignment, Optional<Integer> generation);
```
It's a bit unclear what the compatibility policy should be for plugins, but it seems reasonable to assume that a custom `PartitionAssignor` which does not implement this method will continue to work with the consumer after an upgrade. I think we have to leave the old variant around and add a default implementation which invokes it:
```java
default void onAssignment(Assignment assignment, Optional<Integer> generation) {
  onAssignment(assignment);
}
```
I was debating whether we would want to deprecate the old variant, but it seems not worthwhile since it should be a relatively uncommon case to require the generation.

vahidhashemian replied:
That makes sense. I overlooked the backward compatibility of existing custom assignors.

hachikuji replied:
We still need to add a default implementation which calls `onAssignment` without the generation. Then we can drop the override in `AbstractPartitionAssignor`.

vahidhashemian replied:
That makes sense. Added in the new commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,90 @@
+    public void onAssignment(Assignment assignment, Optional<Integer> generation) {
         memberAssignment = assignment.partitions();
+        this.generation = generation.orElse(DEFAULT_GENERATION);
+        System.out.println("Got " + memberAssignment + " of gen " + this.generation);
```
nit: unintentional? one more in `subscription`

vahidhashemian replied:
Sorry, I missed this cleanup before submitting the patch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -278,7 +279,7 @@ protected void onJoinComplete(int generation,
 
         // give the assignor a chance to update internal state based on the received assignment
-        assignor.onAssignment(assignment);
+        assignor.onAssignment(assignment, Optional.ofNullable(generation));
```
Should be `Optional.of(generation)`. `int` values are not nullable.

vahidhashemian replied:
Good catch, thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -422,10 +503,18 @@ private int getBalanceScore(Map<String, List<TopicPartition>> assignment) {
                 String consumer = sortedConsumers.pollLast();
                 List<TopicPartition> remainingPartitions = assignments.get(consumer);
-                if (!remainingPartitions.isEmpty()) {
+                List<TopicPartition> prevPartitions = new ArrayList<>(prevAssignedPartitions);
```
Do we need `prevAssignedPartitions`? Could we use `prevAssignment.keySet()` here? Some of these operations could probably be more efficient by working sets, but it may not matter too much since this is not a common case.

I'm having a little trouble following the logic here. I find it a little puzzling that we only consider the key set of `prevAssignment`, but I am probably missing something. Perhaps we can add some comments?

vahidhashemian replied:
I added some comments that hopefully clarify the logic a bit. Partitions that were assigned to a different consumer before take priority over those that have stayed with the same consumer. This is to facilitate returning partitions to their previous owner.

And you were right about the parameter switch, only the keys suffice, because for each consumer we basically take all partitions with a previous consumer and consider those that were assigned to each consumer. I hope that makes sense.

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

vahidhashemian committed:
Use `Optional<Integer>` for `generation` to improve backward compatibility

Also, minor fixes and clarifying comments

vahidhashemian reviewed (commented):
@hachikuji I tried to address your comments in the new patch.

Even though I didn't change it, I think using `1` as for `DEFAULT_GENERATION` would be reasonable, as it is the minimum possible generation; just a minor point and it should be fine either way.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -185,24 +187,49 @@
 
     private List<TopicPartition> memberAssignment = null;
     private PartitionMovements partitionMovements;
+    private int generation = DEFAULT_GENERATION; // consumer group generation
```
nit: It might be a little clearer to use `Optional<Integer>` for the type here.

vahidhashemian replied:
I considered following your advice, which probably hints at removing the `DEFAULT_GENERATION`. But that could easily makes things more complex, because at some point(s) we'll have to switch from `Optional` to `int` values because there are sorted structures (`TreeMap`, ...) that rely on default integer comparison. Because of that I feel if we don't want to add complexity we should stay with the `int` type here. Please advise if I misunderstood your point.

hachikuji replied:
Mainly I was just trying to ensure we do not abuse the sentinel, but it sounds like you have thought through its usage.

hachikuji reviewed (commented):
@vahidhashemian Just two comments and I think we can finally merge this.

vahidhashemian committed:
Added a default implementation to the interface

vahidhashemian reviewed (commented):
@hachikuji Thanks for taking another look at this PR. I responded to your comments inline, and submitted another patch.

hachikuji commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -791,6 +792,11 @@ public Subscription subscription(final Set<String> topics) {
     }
 
+    @Override
+    public void onAssignment(final Assignment assignment, final Optional<Integer> generation) {
```
We should be able to get rid of this since we have the default method.

vahidhashemian replied:
True. Removed in the recent patch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -278,38 +305,88 @@
 
     @Override
-    public void onAssignment(Assignment assignment) {
+    public void onAssignment(Assignment assignment, Optional<Integer> generation) {
```
Trying to think through whether generation needs to be an optional. The call inside `ConsumerCoordinator` always passes a valid generation. The only case I can imagine is if an updated assignor gets mixed with an old client. I'm not sure that it's worth the api noise to try and handle that case. What do you think?

vahidhashemian replied:
The impact in the case you mentioned will be minimal at best. I'd also like removing the optional. I've updated the code accordingly.

hachikuji reviewed (commented):
Thanks, just one final api question. Once it is decided, we should update the KIP.

mkedwards commented:
Is this fairly close to being ready to land?  We'll be wanting to use the StickyAssignor when we deploy our latest project to production, but I'm uncomfortable leaving this window open for duplicate partition assignments.  Can I help rebase/test?

vahidhashemian commented:
This is a work in progress on my part. Delayed due to vacation. Will try to resume within a week or so.

hachikuji commented:
@vahidhashemian Do you need any help finishing this up? Would be nice to get it over the line.

vahidhashemian commented:
@hachikuji Sorry for the delay. I'll try to spend time on this over the weekend.

vahidhashemian committed:
Removed the optional qualifier from `onAssignment`'s 'generation` argument

vahidhashemian reviewed (commented):
@hachikuji I submitted another patch to address your latest comments. Let me know what you think. Thanks!

vahidhashemian commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -25,6 +25,7 @@
 import org.apache.kafka.common.errors.GroupMaxSizeReachedException;
 import org.apache.kafka.common.errors.IllegalGenerationException;
 import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.errors.MemberIdRequiredException;
```
This was missed in the rebase.

vahidhashemian commented on clients/src/test/java/org/apache/kafka/clients/consumer/StickyAssignorTest.java:
```diff
@@ -652,7 +652,7 @@ public void testNoExceptionThrownWhenOnlySubscribedTopicDeleted() {
         subscriptions.put(consumer,
                 new Subscription(topics(topic), StickyAssignor.serializeTopicPartitionAssignment(
-                        new ConsumerUserData(assignment.get(consumer), 1))));
+                        new ConsumerUserData(assignment.get(consumer), Optional.of(1)))));
```
This was missed in the rebase.

hachikuji reviewed (approved):
@vahidhashemian LGTM. Thanks for the iterations. Would you mind sending an update to the mail list to mention the differences from the original KIP? We can wait a day or two to make sure there are no objections before merging.
